File: 18-Command/src/cpp/main.cpp
#include <iostream>
#include <memory>
#include <vector>
#include "Light.h"
#include "Command.h"
#include "LightOnCommand.h"
#include "LightOffCommand.h"
#include "MacroCommand.h"
#include "RemoteControl.h"

int main() {
    std::cout << "=== Command Pattern Demo (C++) ===\n\n";

    // Receivers
    Light living_room("Living Room");
    Light kitchen("Kitchen");

    // Concrete commands binding actions to receivers
    auto living_room_on = std::make_shared<LightOnCommand>(living_room);
    auto living_room_off = std::make_shared<LightOffCommand>(living_room);
    auto kitchen_on = std::make_shared<LightOnCommand>(kitchen);

    RemoteControl remote;

    std::cout << "Turn living room light on:\n";
    remote.set_command(living_room_on);
    remote.press_button();

    std::cout << "\nTurn it off:\n";
    remote.set_command(living_room_off);
    remote.press_button();

    std::cout << "\nUndo (should turn back on):\n";
    remote.press_undo();

    std::cout << "\nUndo again (should turn off):\n";
    remote.press_undo();

    // Macro command — "party mode"
    std::cout << "\n--- Macro Command: Party Mode ---\n";
    std::vector<std::shared_ptr<Command>> party = {living_room_on, kitchen_on};
    auto party_mode = std::make_shared<MacroCommand>(party);
    remote.set_command(party_mode);
    remote.press_button();

    std::cout << "\nUndo party mode (reverses both, in reverse order):\n";
    remote.press_undo();

    return 0;
}

File: 18-Command/src/rust/main.rs
mod command;
mod light;
mod light_on_command;
mod light_off_command;
mod macro_command;
mod remote_control;

use std::rc::Rc;

use command::Command;
use light::Light;
use light_on_command::LightOnCommand;
use light_off_command::LightOffCommand;
use macro_command::MacroCommand;
use remote_control::RemoteControl;

fn main() {
    println!("=== Command Pattern Demo (Rust) ===\n");

    // Receivers (shared via Rc so multiple commands can target the same light)
    let living_room = Rc::new(Light::new("Living Room"));
    let kitchen = Rc::new(Light::new("Kitchen"));

    // Concrete commands binding actions to receivers
    let living_room_on: Rc<dyn Command> = Rc::new(LightOnCommand::new(Rc::clone(&living_room)));
    let living_room_off: Rc<dyn Command> = Rc::new(LightOffCommand::new(Rc::clone(&living_room)));
    let kitchen_on: Rc<dyn Command> = Rc::new(LightOnCommand::new(Rc::clone(&kitchen)));

    let mut remote = RemoteControl::new();

    println!("Turn living room light on:");
    remote.set_command(Rc::clone(&living_room_on));
    remote.press_button();

    println!("\nTurn it off:");
    remote.set_command(Rc::clone(&living_room_off));
    remote.press_button();

    println!("\nUndo (should turn back on):");
    remote.press_undo();

    println!("\nUndo again (should turn off):");
    remote.press_undo();

    // Macro command — "party mode"
    println!("\n--- Macro Command: Party Mode ---");
    let party_mode: Rc<dyn Command> =
        Rc::new(MacroCommand::new(vec![Rc::clone(&living_room_on), Rc::clone(&kitchen_on)]));
    remote.set_command(party_mode);
    remote.press_button();

    println!("\nUndo party mode (reverses both, in reverse order):");
    remote.press_undo();
}
